syntax = 'proto3';

package com.science.foundry.tools.grpc_interfaces.pf400;
import "tools/grpc_interfaces/labware.proto";

message Command {
  oneof command {
    Move move = 1;
    GraspPlate grasp_plate = 2;
    ReleasePlate release_plate = 3;
    Approach approach = 4;
    Leave leave = 5;
    Transfer transfer = 6;
    RegisterMotionProfile register_motion_profile = 7;
    SmartTransfer smart_transfer = 8;
    Wait wait = 9;
    Free free = 10;
    UnFree unfree =11;
    Unwind unwind = 12;
    RunSequence run_sequence = 13;
    RetrievePlate retrieve_plate = 14;
    DropOffPlate dropoff_plate = 15;
    PickLid pick_lid = 16;
    PlaceLid place_lid = 17;
    GetCurrentLocation get_current_location = 18;
    Jog jog = 19;
    RawCommand raw_command = 20;
  }

  message RawCommand {
    string command = 1;
  } 

  message GetCurrentLocation {
  }

  message Jog {
    string axis = 1;
    float distance = 2;
  }
  message PickLid {
    string labware = 1;
    string location = 2;
    optional int32 motion_profile_id = 3;
    optional bool pick_from_plate = 4;
  }

  message PlaceLid {
    string labware = 1;
    string location = 2;
    optional int32 motion_profile_id = 3;
    optional bool place_on_plate = 4;
  }

  message RetrievePlate {
    tools.grpc_interfaces.labware.Labware labware = 1;
    string location = 2;
    optional int32 motion_profile_id = 3;
  }

  message DropOffPlate {
    tools.grpc_interfaces.labware.Labware labware = 1;
    string location = 2;
    optional int32 motion_profile_id = 3;
  }

  message RunSequence {
    string sequence_name = 1;
    string labware = 2;
  }
  
  message UnFree {}

  message Free {}

  message Unwind {} 

  message Move {
    string waypoint = 1;
    optional int32 motion_profile_id = 2;
  }

  message GraspPlate {
    int32 width = 1;
    int32 speed = 2;
    int32 force = 3;
  }

  message ReleasePlate {
    int32 width = 1;
    int32 speed = 2;
  }

  message Approach {
    string nest = 1;
    optional float x_offset = 2;
    optional float y_offset = 3;
    optional float z_offset = 4;
    optional int32 motion_profile_id = 5;
    optional string ignore_safepath = 6;
  }

  message Leave {
    string nest = 1;
    optional float x_offset = 2;
    optional float y_offset = 3;
    optional float z_offset = 4;
    optional int32 motion_profile_id = 5;
  }

  message Transfer {
    Approach source_nest = 1;
    Leave destination_nest = 2;
    optional GraspPlate grasp_params = 3;
    optional ReleasePlate release_params = 4;
    optional int32 motion_profile_id = 5;
    optional int32 grip_width = 6;
  }
 
  message RegisterMotionProfile {
    int32 id = 1;
    float speed = 2;
    float speed2 = 3;
    float accel = 4;
    float decel = 5;
    float accel_ramp = 6;
    float decel_ramp = 7;
    float inrange = 8;
    int32 straight = 9;
  }
 
  message SmartTransfer {
    Approach source_nest = 1;
    Leave destination_nest = 2;
    optional GraspPlate grasp_params = 3;
    optional ReleasePlate release_params = 4;
    optional int32 motion_profile_id = 5;
    optional int32 grip_width = 6;
  }

  message Wait {
    int32 duration = 1;
  }

}

message Config {
  string host = 1;
  int32 port = 2;
  int32 joints = 3;
  repeated string waypoints = 4;
}

